### PR TITLE
PCIE MSI and MSI-X multi-vector support plus x86 implementation

### PR DESCRIPTION
**Pull request by tbursztyka**
This PR adds the generic parts for MSI and MSI-X multi-vector support.
Up to each architecture to implement the required arch-specific functions. 
These are already provided in this PR for x86_64.

### CODE REVIEW

**Comment by tbursztyka** [comment]
@andrewboie @andyross @jhedberg @dcpleung please review. Hopefully this could land for the next release.

File: arch/x86/core/pcie.c [hunk 1]
```diff
@@ -155,19 +157,144 @@ void pcie_conf_write(pcie_bdf_t bdf, unsigned int reg, uint32_t data)
...
+
+static const struct device *vtd;
+
+static bool get_vtd(void)
+{
+	if (vtd != NULL) {
+		return true;
+	}
+
+	vtd = device_get_binding("VTD_0");
```
Comment (dcpleung) [inline arch/x86/core/pcie.c:172]: Better to use `DT_*` macro here for device name.

File: include/drivers/pcie/msi.h [hunk 2]
```diff
@@ -7,42 +7,127 @@
...
+ * @brief Allocate vector(s) for the endpoint MSI message(s)
+ * This is arch-specific.
+ *
+ * @param priority the MSI vectors base interrupt priority
+ * @param vectors an array to fill with allocated MSI vectors
+ * @param n_vector the size of MSI vectors array
+ *
+ * @return The number of allocated MSI vectors
+ */
+uint8_t arch_allocate_msi_vectors(unsigned int priority,
```
[the remark this answers is not in the document]
Comment (tbursztyka) [reply include/drivers/pcie/msi.h:50]: sure

File: drivers/pcie/host/msi.c [hunk 3]
```diff
@@ -4,41 +4,122 @@
...
+		pcie_conf_write(bdf, base + PCIE_MSI_MAP1_64, 0U);
+		pcie_conf_write(bdf, base + PCIE_MSI_MDR_64, mdr);
+	} else {
+		pcie_conf_write(bdf, base + PCIE_MSI_MDR_32, mdr);
 	}
 
-	return success;
+	for (i = 0; n_vector > 1; i++) {
+		n_vector >>= 1;
+	}
```
[the remark this answers is not in the document]
Comment (tbursztyka) [reply drivers/pcie/host/msi.c:254]: It's used to generate the right MME value, in 'i' (see the line right below). It's the reverse of what's being done in line 54 (reading MMC field). MMC/MME field are either  0, 1, 2, 4, 8, 16 or 32. I can add a comment there.

File: include/drivers/pcie/msi.h [hunk 4]
```diff
@@ -7,42 +7,127 @@
...
+ * successful.
+ *
+ * @param vector The MSI vector to connect to
+ * @param routine Interrupt service routine
+ * @param parameter ISR parameter
+ * @param flags Arch-specific IRQ configuration flag
+ *
+ * @return True on success, false otherwise
+ */
+bool arch_connect_msi_vector(msi_vector_t *vector,
```
[the remark this answers is not in the document]
Comment (tbursztyka) [reply include/drivers/pcie/msi.h:68]: sure

File: arch/x86/core/pcie.c (hunk 1 above)
Comment (tbursztyka) [reply arch/x86/core/pcie.c:172]: will do

File: include/drivers/pcie/msi.h [hunk 5]
```diff
@@ -7,70 +7,138 @@
...
+ * @param vectors an array to fill with allocated MSI vectors
+ * @param n_vector the size of MSI vectors array
  *
- * Note: PCI(e) capabilities are only used in the MSI code, so for
- * now, capabilities-related code is only included when MSI is. It
- * can easily be separated out if/when its use spreads.
+ * @return The number of allocated MSI vectors
  */
-extern uint32_t pcie_get_cap(pcie_bdf_t bdf, uint32_t cap_id);
+uint8_t arch_allocate_msi_vectors(unsigned int priority,
```
Comment (andrewboie) [inline include/drivers/pcie/msi.h:50]: do not define `arch_` interfaces anywhere other than include/sys/arch_interface.h or kernel/include/kernel_arch_interface.h

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]

**Comment by tbursztyka** [comment]
@andrewboie @dcpleung  please re-review

**Review by tyrantlizard77 — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**